Track fragments on null linked fields

NevilleS opened the pull request:
This fixes #745.

Essentially, what's happening is that the query tracker isn't being updated by `diffRelayQuery` when the query is on a node that is currently `null` in the local store. This causes Relay to lose track of what queries the components are expecting and therefore future mutations form incorrect queries.

My current fix just changes a bit of logic in `diffRelayQuery` to always track null linked fields. I'm not sure it's correct though - I'd like to write some code that ensures the `RelayQueryTracker` correctly merges the two tracked queries that `diffRelayQuery` registers via `trackNodeForID`, but maybe that's already covered by existing tests.

josephsavona commented:
The general approach here feels correct, but there are other places within `diffRelayQuery` where a null link can be reached, and those should be handled similarly (plural linked fields and null connections). It might be helpful to add tests for these cases to find a common place to add the new logic.

> I'd like to write some code that ensures the RelayQueryTracker correctly merges the two tracked queries that diffRelayQuery registers

Tracked queries aren't merged, so no need to test this.

NevilleS commented:
I expected as much. I'll take a look later this week and make some changes.

Does a similar issue exist with the RelayQueryWriter, though? It also calls
into the query tracking logic. When I was poking around it seemed like that
wasn't the case (since that code only deals with what comes back from the
server, not what was asked for in the first place), but I could be wrong.

On Wed, Jan 27, 2016, 6:15 PM Joseph Savona <notifications@github.com>
wrote:

> The general approach here feels correct, but there are other places within
> diffRelayQuery where a null link can be reached, and those should be
> handled similarly (plural linked fields and null connections). It might be
> helpful to add tests for these cases to find a common place to add the new
> logic.
>
> I'd like to write some code that ensures the RelayQueryTracker correctly
> merges the two tracked queries that diffRelayQuery registers
>
> Tracked queries aren't merged, so no need to test this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/relay/pull/774#issuecomment-175906921>.
>

josephsavona commented:
The writer adds tracked queries for newly created nodes, which by definition are non-null, so i don't believe this issue could occur there.

NevilleS commented:
OK, thanks for confirming that particular suspicion :smile: 

I think this issue could theoretically affect any field type that can have a sub-selection (and therefore need fragment tracking). I suppose that only rules out scalar fields, so I'll spend some time writing tests that exercise:

- [X] tracking fragments on null linked fields
- [ ] tracking fragments on null connections
- [ ] tracking fragments on null plural fields
- [ ] tracking fragments on null edges (?)

Does that sound about right?

josephsavona commented:
That sounds right. Note that we don't track queries for records with a "client" id - (ie any record where the server didn't return an `id`) - and edges fall into this category. So we don't need to worry about tracking queries on null edges.

NevilleS commented:
OK, I added some tests this morning for plural fields & connections, and found that the same issue seemed to be present there (unsurprisingly). I've rebased this PR to the latest master and re-pushed. Let me know what to clean up, change, etc.

josephsavona commented:
This is a phenomenal PR - awesome job on tests. Thank you!

josephsavona commented:
@facebook-github-bot import

NevilleS commented:
No problem. Shall I rebase to the latest master?

On Thu, Feb 4, 2016, 11:06 AM Joseph Savona <notifications@github.com>
wrote:

> This is a phenomenal PR - awesome job on tests. Thank you!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/relay/pull/774#issuecomment-179919313>.
>

josephsavona commented:
@NevilleS its ok, our tool can rebase on import.

NevilleS commented:
:open_mouth: We live in the future!
